File: 2017/15.jl
using Base.Threads: @spawn

function generator(num::Int64, gen::Int64, divisor::Int64)
    num = (num * gen) % divisor
    return num 
end

function generator2(num::Int64, gen::Int64, divisor::Int64, mod::Int64)
    while true 
        num = (num * gen) % divisor
        if num % mod == 0
            break
        end
    end
    return num 
end 

function judge(part::Int, n::Int)
    #println(Threads.threadid())
    t = time()
    matches = 0
    na = 703
    nb = 516
    for i in 1:n
        if part == 1
            na = generator(na, 16807, 2147483647)
            nb = generator(nb, 48271, 2147483647)
        else
            na = generator2(na, 16807, 2147483647, 4)
            nb = generator2(nb, 48271, 2147483647, 8)
        end
        
        if (na & 0xFFFF) == (nb & 0xFFFF)
            matches += 1
        end
    end
    println(matches)
    println(time()-t)
end

@sync begin
    @spawn judge(2, 5000000)
    @spawn judge(1, 40000000)
end

File: 2017/15.py
def generator(N, C, M=1):
    while True:
        N = N * C % 2147483647
        if N % M == 0:
            yield N & 0xffff

A, B = 703, 516

gA, gB = generator(A, 16807), generator(B, 48271)
print ("🎄Part 1: {}".format(sum(next(gA) == next(gB) for _ in range(40000000))))

gA, gB = generator(A, 16807, 4), generator(B, 48271, 8)
print ("🎁🎄Part 2: {}".format(sum(next(gA) == next(gB) for _ in range(5000000))))

